Signal/src/view controllers/SignalAttachment.swift

```diff
@@ -0,0 +1,501 @@
+        for dataUTI in videoUTISet {
+            if pasteboardUTISet.contains(dataUTI) {
+                let data = UIPasteboard.general.data(forPasteboardType:dataUTI, inItemSet:itemSet)![0] as! Data
+                return videoAttachment(data : data, dataUTI : dataUTI)
+            }
+        }
+        for dataUTI in audioUTISet {
+            if pasteboardUTISet.contains(dataUTI) {
+                let data = UIPasteboard.general.data(forPasteboardType:dataUTI, inItemSet:itemSet)![0] as! Data
+                return audioAttachment(data : data, dataUTI : dataUTI)
+            }
+        }
+        // TODO: We could handle generic attachments at this point.
+
+        return nil
+    }
+
+    // MARK: Image Attachments
+
+    // Factory method for an image attachment.
+    //
+    // NOTE: The attachment returned by this method may not be valid.
+    //       Check the attachment's error property.
+    public class func imageAttachment(data imageData: Data?, dataUTI: String) -> SignalAttachment {
+        assert(dataUTI.characters.count > 0)
+
+        assert(imageData != nil)
+        guard let imageData = imageData else {
+            let attachment = SignalAttachment(data : Data(), dataUTI: dataUTI)
+            attachment.error = .missingData
+            return attachment
+        }
+
+        let attachment = SignalAttachment(data : imageData, dataUTI: dataUTI)
+
+        guard inputImageUTISet.contains(dataUTI) else {
+            attachment.error = .invalidFileFormat
+            return attachment
+        }
+
+        guard imageData.count > 0 else {
+            assert(imageData.count > 0)
+            attachment.error = .invalidData
+            return attachment
+        }
+
+        if animatedImageUTISet.contains(dataUTI) {
+            guard imageData.count <= kMaxFileSizeAnimatedImage else {
+                attachment.error = .fileSizeTooLarge
+                return attachment
+            }
+            // Never re-encode animated images (i.e. GIFs) as JPEGs.
+            Logger.verbose("\(TAG) Sending raw \(attachment.mimeType) to retain any animation")
+            return attachment
+        } else {
+            guard let image = UIImage(data:imageData) else {
+                attachment.error = .couldNotParseImage
+                return attachment
+            }
+            attachment.image = image
+
+            if isInputImageValidOutputImage(image: image, imageData: imageData, dataUTI: dataUTI) {
+                Logger.verbose("\(TAG) Sending raw \(attachment.mimeType)")
+                return attachment
+            }
+
+            Logger.verbose("\(TAG) Compressing attachment as image/jpeg")
+            return compressImageAsJPEG(image : image, attachment : attachment)
+        }
+    }
+
+    private class func defaultImageUploadQuality() -> TSImageQuality {
+        // Always default to the highest possible image quality and
+        // not converting the image if possible.
+        return .uncropped
```

Comment:
Just to double back, we agreed to change this back so as to conservatively match prod for now, right?